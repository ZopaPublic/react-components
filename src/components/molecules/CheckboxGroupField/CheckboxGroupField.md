### Summary

`<CheckboxGroupField />` is a convenient wrapper that renders group of `<CheckboxField>`s:

⚠️ &nbsp;Note that the `name` for each item has to be unique as it determines the `id`
and `htmlFor` prop for each checkbox field.

### Examples

- Uncontrolled

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'one', label: 'label one' },
    { name: 'two', label: 'label two' },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
/>;
```

- Uncontrolled with one checkbox checked by default

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'three', label: 'label three' },
    { name: 'four', label: 'label four', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
/>;
```

- Uncontrolled with hidden control

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'three', label: 'label three' },
    { name: 'four', label: 'label four', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
  hideControl
/>;
```

- Controlled

```js
import { useState } from 'react';
import { CheckboxGroupField } from '@zopauk/react-components';

const ControlledCheckboxGroupField = () => {
  const [value, setValue] = useState({ five: true, six: false });
  return (
    <CheckboxGroupField
      items={[
        { name: 'five', label: 'label five' },
        { name: 'six', label: 'label six' },
      ]}
      value={value}
      onChange={setValue}
      label="Checkbox group field label"
    />
  );
};

<ControlledCheckboxGroupField />;
```

- Disabled

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'seven', label: 'label seven' },
    { name: 'eight', label: 'label eight' },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
  disabled
/>;
```

- Disabled and valid

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'eight', label: 'label eight' },
    { name: 'nine', label: 'label nine', defaultChecked: true },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
  isValid={true}
  disabled
/>;
```

- With error

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  items={[
    { name: 'ten', label: 'label ten' },
    { name: 'eleven', label: 'label eleven' },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
  errorMessage="You have to check at least one"
/>;
```

- Responsive

```jsx
import { CheckboxGroupField } from '@zopauk/react-components';

<CheckboxGroupField
  flexColProps={{
    xs: 12,
    m: 6,
    xl: 3,
  }}
  items={[
    { name: 'twelve', label: 'label twelve' },
    { name: 'thirteen', label: 'label thirteen' },
    { name: 'fourteen', label: 'label fourteen' },
    { name: 'fifteen', label: 'label fifteen' },
  ]}
  onChange={(value) => console.log(value)}
  label="Checkbox group field label"
/>;
```
