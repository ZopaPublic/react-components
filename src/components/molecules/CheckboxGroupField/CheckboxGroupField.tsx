import React, { useState } from 'react';
import styled from 'styled-components';
import FlexRow, { FlexRowProps } from '../../layout/FlexRow/FlexRow';
import FlexCol, { FlexColProps } from '../../layout/FlexCol/FlexCol';
import CheckboxField from '../CheckboxField/CheckboxField';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import Fieldset from '../../atoms/Fieldset/Fieldset';
import Legend from '../../atoms/Legend/Legend';

const CheckboxWrapper = styled.div`
  padding: 4px 0;
`;

interface CheckboxGroupFieldItem<Val extends Record<string, boolean>> {
  label: string;
  name: keyof Val;
  defaultChecked?: boolean;
  disabled?: boolean;
}

export interface CheckboxGroupFieldProps<Val extends Record<string, boolean>> {
  label: string;
  items: CheckboxGroupFieldItem<Val>[];
  onChange?: (value: Val) => void;
  onBlur?: () => void;
  disabled?: boolean;
  isValid?: boolean;
  value?: Val;
  errorMessage?: string;
  className?: string;
  flexColProps?: FlexColProps;
  flexRowProps?: FlexRowProps;
  hideControl?: boolean;
  'data-automation'?: string;
}

const CheckboxGroupField = <Val extends Record<string, boolean>>({
  items,
  label,
  onChange,
  onBlur,
  value,
  disabled,
  isValid = false,
  errorMessage,
  className,
  flexRowProps = {},
  flexColProps = {},
  hideControl,
  'data-automation': dataAutomation,
}: CheckboxGroupFieldProps<Val>) => {
  const [innerValue, setInnerValue] = useState<Val>(
    items.reduce(
      (acc, { name, defaultChecked }) => ({
        ...acc,
        [name]: defaultChecked || false,
      }),
      {} as Val,
    ),
  );

  const handleChange = (name: keyof Val) => () => {
    if (disabled) {
      return;
    }
    if (!value) {
      const newValue = {
        ...innerValue,
        [name]: !innerValue[name],
      };
      setInnerValue(newValue);
      onChange && onChange(newValue);
    } else {
      onChange &&
        onChange({
          ...value,
          [name]: !value[name],
        });
    }
  };

  return (
    <Fieldset className={className}>
      <Legend>{label}</Legend>
      <FlexRow {...flexRowProps}>
        {items.map((item, index) => {
          const checked = value ? !!value[item.name] : !!innerValue[item.name];
          return (
            <FlexCol {...flexColProps} key={item.name.toString()}>
              <CheckboxWrapper>
                <CheckboxField
                  name={item.name.toString()}
                  disabled={disabled}
                  onChange={handleChange(item.name)}
                  onBlur={onBlur}
                  label={item.label}
                  checked={checked}
                  isValid={checked && isValid}
                  hideControl={hideControl}
                  data-automation={dataAutomation ? `${dataAutomation}-${index}` : undefined}
                />
              </CheckboxWrapper>
            </FlexCol>
          );
        })}
      </FlexRow>
      {errorMessage && <ErrorMessage className="mt-1">{errorMessage}</ErrorMessage>}
    </Fieldset>
  );
};

export default CheckboxGroupField;
