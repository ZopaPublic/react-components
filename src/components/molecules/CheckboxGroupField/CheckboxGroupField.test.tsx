import React, { useState } from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import axe from '../../../../axe-helper';
import CheckboxGroupField from './CheckboxGroupField';

describe('<CheckboxGroupField />', () => {
  it('should handle value change if it is not controlled', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <CheckboxGroupField
        label="label"
        onChange={onChange}
        items={[
          {
            name: 'one',
            label: 'label one',
            defaultChecked: true,
          },
          {
            name: 'two',
            label: 'label two',
          },
        ]}
      />,
    );
    const labelOne = getByLabelText('label one');
    const labelTwo = getByLabelText('label two');
    if (labelOne instanceof HTMLInputElement && labelTwo instanceof HTMLInputElement) {
      expect(labelOne.checked).toEqual(true);
      expect(labelTwo.checked).toEqual(false);

      fireEvent.click(labelTwo);
      expect(onChange).toHaveBeenCalledWith({ one: true, two: true });
      expect(labelOne.checked).toEqual(true);
      expect(labelTwo.checked).toEqual(true);
    } else {
      throw new Error('expected labelOne and labelTwo to be html input elements');
    }
  });

  it('should not change value if it is disabled', () => {
    const onChange = jest.fn();
    const { getByLabelText } = render(
      <CheckboxGroupField
        disabled
        label="label"
        onChange={onChange}
        items={[
          {
            name: 'one',
            label: 'label one',
          },
          {
            name: 'two',
            label: 'label two',
          },
        ]}
      />,
    );
    const labelOne = getByLabelText('label one');
    const labelTwo = getByLabelText('label two');

    if (labelOne instanceof HTMLInputElement && labelTwo instanceof HTMLInputElement) {
      expect(labelOne.checked).toEqual(false);
      expect(labelTwo.checked).toEqual(false);
      fireEvent.click(labelOne);
      expect(onChange).not.toHaveBeenCalledWith('one');
      expect(labelOne.checked).toEqual(false);
      expect(labelOne.checked).toEqual(false);

      fireEvent.click(labelTwo);
      expect(onChange).not.toHaveBeenCalledWith('two');
      expect(labelOne.checked).toEqual(false);
      expect(labelTwo.checked).toEqual(false);
    } else {
      throw new Error('expected labelOne and labelTwo to be html input elements');
    }
  });

  it('should handle value change if it is controlled', () => {
    const ControlledCheckboxGroupField = () => {
      const [value, setValue] = useState<Record<string, boolean>>({ one: true, two: false });
      return (
        <CheckboxGroupField
          items={[
            { name: 'one', label: 'label one' },
            { name: 'two', label: 'label two' },
          ]}
          value={value}
          onChange={setValue}
          label="label"
        />
      );
    };
    const { getByLabelText } = render(<ControlledCheckboxGroupField />);
    const labelOne = getByLabelText('label one');
    const labelTwo = getByLabelText('label two');
    if (labelOne instanceof HTMLInputElement && labelTwo instanceof HTMLInputElement) {
      expect(labelOne.checked).toEqual(true);
      expect(labelTwo.checked).toEqual(false);
      act(() => {
        fireEvent.click(labelTwo);
      });
      expect(labelOne.checked).toEqual(true);
      expect(labelTwo.checked).toEqual(true);
    } else {
      throw new Error('expected labelOne and labelTwo to be html input elements');
    }
  });

  it('renders the component with props with no a11y violations', async () => {
    const { container } = render(
      <CheckboxGroupField
        label="label"
        onChange={jest.fn()}
        items={[{ label: 'option', name: 'option' }]}
        data-automation="ZA.test"
      />,
    );
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });

  it('renders the component with no control icon and no a11y violations', async () => {
    const { container } = render(
      <CheckboxGroupField
        label="label"
        onChange={jest.fn()}
        items={[{ label: 'option', name: 'option' }]}
        hideControl
      />,
    );
    const results = await axe(container.innerHTML);

    expect(container.firstChild).toMatchSnapshot();
    expect(results).toHaveNoViolations();
  });
});
