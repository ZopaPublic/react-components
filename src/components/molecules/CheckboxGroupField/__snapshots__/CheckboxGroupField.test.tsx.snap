// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxGroupField /> renders the component with no control icon and no a11y violations 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c7 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c5 {
  max-width: 100%;
  width: 100%;
}

.c8 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  font-weight: 400;
  font-size: 16px;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  line-height: 1.4;
  color: #2C3236;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0;
}

.c8:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 4px;
  height: 20px;
  width: 20px;
  margin-right: 8px;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  box-shadow: 0 0 4px 0 transparent;
  border: 1px #818F9B solid;
  display: block;
}

.c8:after {
  content: '';
  background-size: contain;
  background: transparent none no-repeat center;
  height: 12px;
  width: 12px;
  position: absolute;
  left: 20px;
  top: 18px;
}

.c8:hover {
  cursor: pointer;
}

.c6 {
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c6:hover:not(:disabled) + label,
.c6:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c6:hover:not(:disabled) + label:before,
.c6:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c6:hover:checked:not(:disabled) + label:after,
.c6:focus:checked + label:after {
  background-image: url(teal-check-mark.svg);
}

.c6:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c6:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c6:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c6:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c6:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c6:checked + label:after {
  background-size: contain;
  background-image: url(teal-check-mark.svg);
  -webkit-animation: guYtTV 180ms ease-in-out;
  animation: guYtTV 180ms ease-in-out;
}

.c6 + label:before,
.c6 + label:after {
  display: none;
}

.c0 {
  border: 0;
  padding: 0.01em 0 0 0;
  margin: 0;
  min-width: 0;
}

.c1 {
  padding: 0;
  display: table;
  margin-bottom: 8px;
  font-weight: 600;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c4 {
  padding: 4px 0;
}

<fieldset
  class="c0"
>
  <legend
    class="c1"
  >
    label
  </legend>
  <div
    class="c2"
    cols="12"
    direction="row"
  >
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <input
            class="c6"
            id="checkbox-id-option"
            name="option"
            type="checkbox"
          />
          <label
            class="c7 c8"
            for="checkbox-id-option"
          >
            option
          </label>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;

exports[`<CheckboxGroupField /> renders the component with props with no a11y violations 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin-right: -12px;
  margin-left: -12px;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c3 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: 12px;
  padding-left: 12px;
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
}

.c7 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c5 {
  max-width: 100%;
  width: 100%;
}

.c8 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  text-align: left;
  font-weight: 400;
  font-size: 16px;
  padding: 14px 16px;
  border: 1px solid #818F9B;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  border-radius: 8px;
  line-height: 1.4;
  color: #2C3236;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  margin-bottom: 0;
}

.c8:before {
  content: '';
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  background-color: #FFFFFF;
  border-radius: 4px;
  height: 20px;
  width: 20px;
  margin-right: 8px;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
  box-shadow: 0 0 4px 0 transparent;
  border: 1px #818F9B solid;
  display: block;
}

.c8:after {
  content: '';
  background-size: contain;
  background: transparent none no-repeat center;
  height: 12px;
  width: 12px;
  position: absolute;
  left: 20px;
  top: 18px;
}

.c8:hover {
  cursor: pointer;
}

.c6 {
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c6:hover:not(:disabled) + label,
.c6:focus + label {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c6:hover:not(:disabled) + label:before,
.c6:focus + label:before {
  border-color: #009E8E;
  box-shadow: 0 0 4px 0 #009E8E;
}

.c6:hover:checked:not(:disabled) + label:after,
.c6:focus:checked + label:after {
  background-image: url(teal-check-mark.svg);
}

.c6:disabled + label {
  cursor: not-allowed;
  color: #818F9B;
  background-color: #F7F7F7;
}

.c6:disabled:not(:checked) + label {
  border-color: #D4D7D9;
}

.c6:disabled:not(:checked) + label:before {
  border-color: #D4D7D9;
}

.c6:checked + label {
  border-color: #009E8E;
  background-color: #ECFCF9;
  box-shadow: none;
}

.c6:checked + label:before {
  border-color: #009E8E;
  box-shadow: none;
}

.c6:checked + label:after {
  background-size: contain;
  background-image: url(teal-check-mark.svg);
  -webkit-animation: guYtTV 180ms ease-in-out;
  animation: guYtTV 180ms ease-in-out;
}

.c0 {
  border: 0;
  padding: 0.01em 0 0 0;
  margin: 0;
  min-width: 0;
}

.c1 {
  padding: 0;
  display: table;
  margin-bottom: 8px;
  font-weight: 600;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
}

.c4 {
  padding: 4px 0;
}

<fieldset
  class="c0"
>
  <legend
    class="c1"
  >
    label
  </legend>
  <div
    class="c2"
    cols="12"
    direction="row"
  >
    <div
      class="c3"
      cols="12"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <input
            class="c6"
            data-automation="ZA.test-0"
            id="checkbox-id-option"
            name="option"
            type="checkbox"
          />
          <label
            class="c7 c8"
            for="checkbox-id-option"
          >
            option
          </label>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
